#Directory, Libraries and Data

In [1]:
%cd /content/drive/MyDrive/Business Analyst course/Econometrics and Causal Inference/Matching

/content/drive/MyDrive/Business Analyst course/Econometrics and Causal Inference/Matching


In [2]:
#install library
!pip install CausalInference

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.1/51.1 kB 1.3 MB/s eta 0:00:00


In [3]:
#Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import scipy.stats as ss
from causalinference import CausalModel

In [5]:
#Loading Data
df = pd.read_csv("stackoverflow.csv")
df.head()

Country         Salary  YearsCodedJob  OpenSource  Hobby  \
0  United Kingdom  100000.000000             20           0      1   
1   United States  130000.000000             20           1      1   
2   United States  175000.000000             16           0      1   
3         Germany   64516.129030              4           0      0   
4           India    6636.323594              1           0      1   

   CompanySizeNumber      Remote  CareerSatisfaction  Data_scientist  \
0               5000      Remote                   8               0   
1               1000      Remote                   9               0   
2              10000  Not remote                   7               0   
3               1000  Not remote                   9               0   
4               5000  Not remote                   5               0   

   Database_administrator  ...  Developer_with_stats_math_background  DevOps  \
0                       0  ...                                     0       0   
1                       0  ...                                     0       1   
2                       0  ...                                     0       0   
3                       0  ...                                     0       0   
4                       0  ...                                     0       0   

   Embedded_developer  Graphic_designer  Graphics_programming  \
0                   1                 0                     0   
1                   1                 0                     0   
2                   0                 0                     0   
3                   0                 0                     0   
4                   0                 0                     0   

   Machine_learning_specialist  Mobile_developer  Quality_assurance_engineer  \
0                            0                 0                           0   
1                            0                 0                           1   
2                            0                 0                           0   
3                            0                 0                           0   
4                            0                 0                           0   

   Systems_administrator  Web_developer  
0                      0              0  
1                      0              1  
2                      0              1  
3                      0              1  
4                      0              1  

[5 rows x 21 columns]

#Data Analysis

In [6]:
#Picking variables
df = df.iloc[:,:8]
df.head(0)

Empty DataFrame
Columns: [Country, Salary, YearsCodedJob, OpenSource, Hobby, CompanySizeNumber, Remote, CareerSatisfaction]
Index: []

In [7]:
#Transforming character variables
df = pd.get_dummies(df, drop_first = True)
df.head(1)

Salary  YearsCodedJob  OpenSource  Hobby  CompanySizeNumber  \
0  100000.0             20           0      1               5000   

   CareerSatisfaction  Country_Germany  Country_India  Country_United Kingdom  \
0                   8                0              0                       1   

   Country_United States  Remote_Remote  
0                      0              1

In [8]:
#Looking at Group Means
df.groupby("Remote_Remote").mean()

Salary  YearsCodedJob  OpenSource     Hobby  \
Remote_Remote                                                      
0              70201.175042       7.142857    0.332736  0.760510   
1              87400.737001      10.120000    0.443478  0.766957   

               CompanySizeNumber  CareerSatisfaction  Country_Germany  \
Remote_Remote                                                           
0                    2187.325563            7.551106         0.142857   
1                    1712.756522            7.855652         0.069565   

               Country_India  Country_United Kingdom  Country_United States  
Remote_Remote                                                                
0                   0.096035                0.189878               0.480175  
1                   0.097391                0.121739               0.662609

In [10]:
#T-tests
#stating with variables to test
continuous = ['Salary', 'YearsCodedJob']

#Where to store results
stat = {}
pvalue = {}

#Loop
for x in continuous:
  group1 = df.where(df.Remote_Remote == 0).dropna()[x]
  group2 = df.where(df.Remote_Remote == 1).dropna()[x]
  stat[x], pvalue[x] = ss.ttest_ind(group1, group2)
ttests = pd.DataFrame.from_dict(pvalue, orient = 'Index')
ttests.columns = ['pvalue']
print(ttests)

                     pvalue
Salary         1.057708e-22
YearsCodedJob  3.637316e-30


#Matching

In [11]:
#Isolate y, treat, confounders
y = df.CareerSatisfaction.values
treat = df.Remote_Remote.values
confounders = df.drop(columns = ["Remote_Remote",
                                 "CareerSatisfaction"]).values

In [12]:
#Matching
model = CausalModel(y, treat, confounders)
model.est_via_matching(bias_adj = True)
print(model.estimates)

/usr/local/lib/python3.10/dist-packages/causalinference/estimators/matching.py:100: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  return np.linalg.lstsq(X, Y)[0][1:]  # don't need intercept coef



Treatment Effect Estimates: Matching

                     Est.       S.e.          z      P>|z|      [95% Conf. int.]
--------------------------------------------------------------------------------
           ATE      0.114      0.130      0.881      0.378     -0.140      0.368
           ATC      0.106      0.136      0.778      0.436     -0.161      0.372
           ATT      0.187      0.142      1.321      0.187     -0.090      0.464



ATE (Average Treatment Effect): The estimated treatment effect size for the entire treated group is approximately 0.114, but the standard error is quite high (0.130). The 'z' score (0.881) suggests that this effect is not statistically significant (p-value = 0.378). The 95% confidence interval ranges from -0.140 to 0.368, indicating a wide range of potential effects.

ATC (Average Treatment Effect on the Control Group): This represents the effect of the treatment on the control group. The estimated effect size is 0.106, but like the ATE, it's not statistically significant (z = 0.778, p-value = 0.436). The confidence interval spans from -0.161 to 0.372.

ATT (Average Treatment Effect on the Treated): This shows the effect of the treatment specifically on the group that received the treatment. The estimated effect size is 0.187, but again, it's not statistically significant (z = 1.321, p-value = 0.187). The confidence interval ranges from -0.090 to 0.464.

#Robustness Check

In [13]:
#Remove 1 confounder
confounders = df.drop(columns = ["Remote_Remote",
                                 "CareerSatisfaction",
                                 "Hobby"]).values

In [14]:
#Matching
model = CausalModel(y, treat, confounders)
model.est_via_matching(bias_adj = True)
print(model.estimates)

/usr/local/lib/python3.10/dist-packages/causalinference/estimators/matching.py:100: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  return np.linalg.lstsq(X, Y)[0][1:]  # don't need intercept coef



Treatment Effect Estimates: Matching

                     Est.       S.e.          z      P>|z|      [95% Conf. int.]
--------------------------------------------------------------------------------
           ATE      0.148      0.131      1.128      0.259     -0.109      0.406
           ATC      0.140      0.138      1.013      0.311     -0.131      0.411
           ATT      0.220      0.137      1.602      0.109     -0.049      0.488



ATE (Average Treatment Effect): The estimated treatment effect size for the entire treated group is approximately 0.148. The standard error is 0.131, indicating some uncertainty. The 'z' score (1.128) suggests that this effect is not statistically significant (p-value = 0.259). The 95% confidence interval ranges from -0.109 to 0.406.

ATC (Average Treatment Effect on the Control Group): This represents the effect of the treatment on the control group. The estimated effect size is 0.140, with a standard error of 0.138. The 'z' score (1.013) suggests that this effect is also not statistically significant (p-value = 0.311). The confidence interval spans from -0.131 to 0.411.

ATT (Average Treatment Effect on the Treated): This shows the effect of the treatment specifically on the group that received the treatment. The estimated effect size is 0.220, with a standard error of 0.137. The 'z' score (1.602) suggests that this effect is not statistically significant either (p-value = 0.109). The confidence interval ranges from -0.049 to 0.488.

In this robustness check, it appears that the treatment effects are not statistically significant for any of the groups (ATE, ATC, or ATT). This means that the observed effects may not be reliable or may not be due to the treatment itself. Researchers use robustness checks to see if their results hold up under different conditions, and in this case, the effects remain inconclusive in terms of statistical significance.